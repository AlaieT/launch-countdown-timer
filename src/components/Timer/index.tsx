import React, { useEffect, useState } from "react";

import Digit from "../Digit";

import type { TimerProps } from "../../types";

const SECOND = 1000;
const MINUTE = SECOND * 60;
const HOUR = MINUTE * 60;
const DAY = HOUR * 24;

const leftTime = (deadline: string): number => {
  const diff = Date.parse(deadline) - Date.now();
  return diff > 0 ? diff : 0;
};

function Timer({ deadline }: TimerProps) {
  const [time, setTime] = useState(leftTime(deadline));

  useEffect(() => {
    const interval = setInterval(() => {
      setTime(leftTime(deadline));
    }, 1000);

    return () => clearInterval(interval);
  }, [deadline]);

  return (
    <>
      <Digit label="days" digit={Math.floor(time / DAY)} />
      <Digit label="hours" digit={Math.floor((time / HOUR) % 24)} />
      <Digit label="minuts" digit={Math.floor((time / MINUTE) % 60)} />
      <Digit label="seconds" digit={Math.floor((time / SECOND) % 60)} />
    </>
  );
}

export default Timer;
