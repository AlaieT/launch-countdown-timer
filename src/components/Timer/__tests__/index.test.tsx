import React from "react";
import { render } from "@testing-library/react";

import Timer from "..";

describe("Timer testing...", () => {
  const props = {
    deadline: "April 30, 2023"
  };

  it("Snapshot testing", () => {
    const { asFragment } = render(<Timer deadline={props.deadline} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
