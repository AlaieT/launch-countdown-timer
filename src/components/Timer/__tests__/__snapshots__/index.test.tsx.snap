// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer testing... Snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="digit"
  >
    <div
      class="box"
    >
      <div
        class="top"
      >
        <p>
          21
        </p>
      </div>
      <div
        class="middle"
      >
        <div
          class="front"
        >
          <p>
            21
          </p>
        </div>
        <div
          class="back"
        >
          <p>
            21
          </p>
        </div>
      </div>
      <div
        class="bottom"
      >
        <p>
          21
        </p>
      </div>
    </div>
    <div
      class="label"
    >
      <p>
        days
      </p>
    </div>
  </div>
  <div
    class="digit"
  >
    <div
      class="box"
    >
      <div
        class="top"
      >
        <p>
          18
        </p>
      </div>
      <div
        class="middle"
      >
        <div
          class="front"
        >
          <p>
            18
          </p>
        </div>
        <div
          class="back"
        >
          <p>
            18
          </p>
        </div>
      </div>
      <div
        class="bottom"
      >
        <p>
          18
        </p>
      </div>
    </div>
    <div
      class="label"
    >
      <p>
        hours
      </p>
    </div>
  </div>
  <div
    class="digit"
  >
    <div
      class="box"
    >
      <div
        class="top"
      >
        <p>
          19
        </p>
      </div>
      <div
        class="middle"
      >
        <div
          class="front"
        >
          <p>
            19
          </p>
        </div>
        <div
          class="back"
        >
          <p>
            19
          </p>
        </div>
      </div>
      <div
        class="bottom"
      >
        <p>
          19
        </p>
      </div>
    </div>
    <div
      class="label"
    >
      <p>
        minuts
      </p>
    </div>
  </div>
  <div
    class="digit"
  >
    <div
      class="box"
    >
      <div
        class="top"
      >
        <p>
          07
        </p>
      </div>
      <div
        class="middle"
      >
        <div
          class="front"
        >
          <p>
            07
          </p>
        </div>
        <div
          class="back"
        >
          <p>
            07
          </p>
        </div>
      </div>
      <div
        class="bottom"
      >
        <p>
          07
        </p>
      </div>
    </div>
    <div
      class="label"
    >
      <p>
        seconds
      </p>
    </div>
  </div>
</DocumentFragment>
`;
